Add text-variable-anchor-offset property

drwestco opened the pull request:
## Launch Checklist

<!-- Thanks for the PR! Feel free to add or remove items from the checklist. -->
 
Per discussion and voting in issue #112 , added text-variable-anchor-offset symbol layer layout property to the style spec.

 - [X] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [X] Briefly describe the changes in this PR.
 - [X] Link to related issues.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
 - [X] Write tests for all new functionality.
 - [X] Document any changes to public APIs.
 - [ ] Post benchmark scores.
 - [X] Add an entry to `CHANGELOG.md` under the `## main` section.

drwestco commented on src/reference/v8.json:
```diff
@@ -2067,6 +2067,30 @@
+          "zoom"
+        ]
+      },
+      "property-type": "data-constant"
```
`text-variable-anchor` is `data-constant`, but `text-offset` is `data-driven`. Since this is a superset of both properties, maybe `data-driven` is more appropriate?

HarelM replied:
Sounds right.

HarelM reviewed (approved):
This looks great! Thanks!!

HarelM commented:
Did you have a change to test this with maplibre-gl? Can you share a video of the test? I'm supper intrigued to see how this works :-)

drwestco commented:
> Did you have a change to test this with maplibre-gl?

Not yet. That's a little more complicated and I'm just getting started figuring it out.

HarelM commented:
Is it ok if I wait with merging this to see that this can be used end to end?

drwestco commented:
Sure. I don't yet need it published to npm, so fine to hold off on the merge.

drwestco commented:
Got it working with maplibre-gl-js. Modified the variable-label-placement example as a test. Bottom and right anchors are offset twice as much as top and left.

![image](https://github.com/maplibre/maplibre-style-spec/assets/743030/f13f6e4a-df34-4569-b689-08f55cceee1d)

Implementation in maplibre-gl-js isn't ready to review yet, but I think the style changes here are good. Version support and/or other parts of the documentation may need tweaking?

HarelM commented:
Can you share your steps somewhere? @zelonewolf was trying to achieve the same (using soec with naplibre-gl)

drwestco commented:
I have a thread on Slack with the steps I used. In the style-spec repo, I run the commands
```
npm run generate-style-spec
npm run generate-typings
npm run build
npm pack
```
I copy the resulting .tgz to maplibre-gl, then update style spec dependency in package.json over there
```
"@maplibre/maplibre-gl-style-spec": "file:maplibre-maplibre-gl-style-spec-19.3.0.tgz",
```

ZeLonewolf commented:
This is the approach I've been using to test as well. However, I'm running into issues with transitive dependencies, most notably on an old version of mapbox-gl and its embedded style spec stomping on the separately-built style-spec.

HarelM commented:
Should I merge and release a version then? Or do you need more time to make sure nothing else is needed in this package?

HarelM commented:
I've recently added code coverage to this package, can you rebase and see that the added code is covered in tests?

drwestco commented:
I pushed the rebased changes. How do I verify the code coverage?

HarelM commented:
Hmm, seems like the report wasn't uploaded well, not sure why, can you run the coverage locally to see that the changes are covered with tests?

drwestco commented:
> Hmm, seems like the report wasn't uploaded well, not sure why, can you run the coverage locally to see that the changes are covered with tests?

How do I do that, and what am I looking for?

HarelM commented:
To see that the code that was added is covered with tests.
The CI is running coverage, but there's a need to add report for text, I'll send the command later on.

HarelM commented:
Try the following:
`npm run jest-ci -- --coverage --coverageDirectory=coverage-unit --coverageReporters text --selectProjects unit`
`npm run jest-ci -- --coverage --coverageDirectory=coverage-integration --coverageReporters text --selectProjects integration`

drwestco commented:
Cool. Added a couple of lines to unit test to hit 100% coverage for variable_anchor_offset_collection. The other lines I added to the remaining files were already covered.

HarelM commented:
THANKS!
Can you open up a PR with version number change so that I'll be able to merge it and create a new release?

HarelM commented:
@drwestco are you planning a PR to the main library itself beside a version bump of the style spec package?
I'm asking since I want to release a version soon to update the docs and wanted to know if you want this inside as well...

drwestco commented:
Yes. I have a working implementation that needs a bit more polish, but hopefully can have it ready for review in a week or so.

HarelM commented:
Ok, so I won't wait with releasing a new version, we can release a new one later on.
